### PR TITLE
Implement enhancements/oc/mirroring-release-signatures

### PR DESCRIPTION
**Pull request by jottofar**
This PR implements enhancement https://github.com/openshift/enhancements/pull/283

Local package pkg/verify was reused with minor changes from CVO so there is PR https://github.com/openshift/library-go/pull/725 to move package verify into library-go.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -36,7 +36,7 @@ require (
 	github.com/openshift/api v0.0.0-20200330134433-8e259f67fc55
 	github.com/openshift/build-machinery-go v0.0.0-20200211121458-5e3d6e570160
 	github.com/openshift/client-go v0.0.0-20200326155132-2a6cd50aedd0
-	github.com/openshift/library-go v0.0.0-20200327125526-163b2f0d6264
+	github.com/openshift/library-go v0.0.0-20200402123743-4015ba624cae
```
Comment (wking) [inline go.mod:39]: not sure how the `oc` maintainers like to handle this, but personally I'd rather have the vendor bump and the completion-changes (both auto-generated) pulled out into separate commits to make it easier to focus review on the human-generated code changes.

File: pkg/cli/admin/release/mirror.go [hunk 2]
```diff
@@ -213,6 +370,22 @@ func (o *MirrorOptions) Run() error {
 		return fmt.Errorf("--skip-release-image and --to-release-image may not both be specified")
 	}
 
+	if o.ApplyReleaseImageSignature {
+		if len(o.ReleaseImageSignatureToDir) != 0 {
+			return fmt.Errorf("--release-image-signature-to-dir and --apply-release-image-signature may not both be specified")
```
Comment (LalatenduMohanty) [inline pkg/cli/admin/release/mirror.go:375]: `may not` is not great way to communicate that the flags are not compatible with each other.  we should use `must  not` or some explicit way to communicate.

File: pkg/cli/admin/release/mirror.go [hunk 3]
```diff
@@ -213,6 +370,22 @@ func (o *MirrorOptions) Run() error {
...
+		}
+	} else {
+		if len(o.ReleaseImageSignatureToDir) == 0 {
+			if len(o.ToDir) == 0 {
+				return fmt.Errorf("if --to-dir and --apply-release-image-signature are not specified must use --release-image-signature-to-dir to specify a directory to export signature to")
```
Comment (LalatenduMohanty) [inline pkg/cli/admin/release/mirror.go:380]: change 
`if --to-dir and --apply-release-image-signature are not specified must use --release-image-signature-to-dir to specify a directory to export signature to`  with

 `if --to-dir and --apply-release-image-signature are not specified, --release-image-signature-to-dir must be use to specify a directory to export the signature`

Comment (jottofar) [reply pkg/cli/admin/release/mirror.go:380]: I 100% agree but followed the established pattern in the code to be consistent. Not only in mirror but throughout oc it uses "may not" in it's error output.

File: pkg/cli/admin/release/mirror.go (hunk 2 above)
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go:375]: I 100% agree but followed the established pattern in the code to be consistent. Not only in mirror but throughout oc it uses "may not" in it's error output.

File: pkg/cli/admin/release/extract.go [hunk 4]
```diff
@@ -196,23 +210,77 @@ func (o *ExtractOptions) Run() error {
...
+							manifestErrs = append(manifestErrs, errors.Wrapf(err, "error reading file %s", hdr.Name))
+						}
+						ms, err := manifest.ParseManifests(bytes.NewReader(raw))
+						if err != nil {
+							manifestErrs = append(manifestErrs, errors.Wrapf(err, "error parsing %s", hdr.Name))
+						}
+						for i := range ms {
+							ms[i].OriginalFilename = filepath.Base(hdr.Name)
+							src := fmt.Sprintf("the config map %s/%s", ms[i].Obj.GetNamespace(), ms[i].Obj.GetName())
+							data, _, err := unstructured.NestedStringMap(ms[i].Obj.Object, "data")
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:250]: I reused CVO code wherever possible. Eventually the hope is to move some of this into library-go and share between oc and CVO.

https://github.com/openshift/cluster-version-operator/blob/914467c03b5bc25dbe4c9295c420fdb2598db821/pkg/cvo/cvo.go#L304

File: vendor/github.com/openshift/library-go/pkg/manifest/manifest.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
+package manifest
```
[the remark this answers is not in the document]
Comment (jottofar) [reply vendor/github.com/openshift/library-go/pkg/manifest/manifest.go:1]: How do I handle the chicken-and the-egg issue here. I bumped libray-go first before I added the dependency on manifest package to my code therefore the vendoring did not pull down manifest package. Then I modified the oc code and added the manifest dependency.

File: pkg/verify/configmap.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+// GetSignaturesAsConfigmapBytes stores the given signatures as a configmap and
+// and returns configmap as bytes. Uses digestToKeyPrefix to replace colon with
+// dash when saving digest to configmap.
+func GetSignaturesAsConfigmapBytes(digest string, signatures [][]byte) ([]byte, error) {
+	var cm *corev1.ConfigMap
+	cm, err := GetSignaturesAsConfigmap(digest, signatures)
+	if err != nil {
+		return nil, err
+	}
+	if cmData, err := yaml.Marshal(cm); err == nil {
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go:83]: Will update to use proper k8s decoding in subsequent PR.

File: pkg/cli/admin/release/extract.go [hunk 7]
```diff
@@ -196,23 +210,77 @@ func (o *ExtractOptions) Run() error {
...
+				case "image-references":
+					return true, nil
+				case "release-metadata":
+					return true, nil
+				default:
+					if ext := path.Ext(hdr.Name); len(ext) > 0 && (ext == ".yaml" || ext == ".yml" || ext == ".json") {
+						klog.V(4).Infof("Found manifest %s", hdr.Name)
+						raw, err := ioutil.ReadAll(r)
+						if err != nil {
+							manifestErrs = append(manifestErrs, errors.Wrapf(err, "error reading file %s", hdr.Name))
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:241]: I go through all the manifests and gather any errors encountered. I only care about the two signature manifest files and I've noticed in testing that some of the other manifest files generate errors. After going through all manifests errors will be output if the signature manifests were not found since one or more of the errors will most likely apply. If the signature manifests are found I do not print out the errors since they should be irrelivant and will certainly cause confusion and alarm to the user.

File: pkg/cli/admin/release/extract.go [hunk 8]
```diff
@@ -196,23 +210,77 @@ func (o *ExtractOptions) Run() error {
...
+						}
+						ms, err := manifest.ParseManifests(bytes.NewReader(raw))
+						if err != nil {
+							manifestErrs = append(manifestErrs, errors.Wrapf(err, "error parsing %s", hdr.Name))
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:245]: See https://github.com/openshift/oc/pull/343#discussion_r410361342

File: pkg/verify/configmap.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
+package verify
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go]: Changed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go (hunk 3 above)
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Changed in commit e3f50c1

File: pkg/cli/admin/release/extract.go [hunk 10]
```diff
@@ -196,23 +210,77 @@ func (o *ExtractOptions) Run() error {
...
+							if err != nil {
+								manifestErrs = append(manifestErrs, errors.Wrapf(err, "%s is not valid", src))
+							}
+							for k, v := range data {
+								switch {
+								case strings.HasPrefix(k, "verifier-public-key-"):
+									klog.V(2).Infof("Found in %s:\n%s %s", hdr.Name, k, v)
+								case strings.HasPrefix(k, "store-"):
+									klog.V(2).Infof("Found in %s:\n%s\n%s", hdr.Name, k, v)
+								default:
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/extract.go [hunk 11]
```diff
@@ -196,23 +210,77 @@ func (o *ExtractOptions) Run() error {
...
 		}
 		if !found {
 			return fmt.Errorf("image did not contain %s", o.File)
 		}
+		// Warn user and print errors if any but do not return an error so current
+		// operation, e.g. mirroring, continues
+		if o.ExtractManifests && len(o.Manifests) == 0 {
+			fmt.Fprintf(o.ErrOut, "No manifests found\n")
+			if len(manifestErrs) > 0 {
+				fmt.Fprintf(o.ErrOut, "Errors: %s\n", errorList(manifestErrs))
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:283]: Related to comment https://github.com/openshift/oc/pull/343#discussion_r410361342.

I updated the comment in the code to better explain the logic. 

I go through all the manifests and gather any errors encountered. I only care about the two signature manifest files and I've noticed in testing that some of the other manifest files generate errors. After going through all manifests errors will be output if the signature manifests were not found since one or more of the errors will most likely apply. If the signature manifests are found I do not print out the errors since they should be irrelivant and will certainly cause confusion and alarm to the user.

File: pkg/cli/admin/release/mirror.go [hunk 12]
```diff
@@ -131,10 +173,15 @@ type MirrorOptions struct {
 	ToMirror bool
 	ToDir    string
 
+	ApplyReleaseImageSignature bool
+	ReleaseImageSignatureToDir string
+	OverWrite                  bool
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go [hunk 13]
```diff
@@ -131,10 +173,15 @@ type MirrorOptions struct {
...
+
 	DryRun                        bool
 	PrintImageContentInstructions bool
 
-	ClientFn func() (imageclient.Interface, string, error)
+	ClientFn     func() (imageclient.Interface, string, error)
+	KubeClientFn func() (kubeclient.ConfigMapInterface, error)
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go [hunk 14]
```diff
@@ -176,13 +223,123 @@ func (o *MirrorOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []
...
+	}
+	return fmt.Sprintf(signatureFileNameFmt, algo, hash), nil
+}
+
+// handleSignatures implements the image release signature configmap specific logic.
+// Signature configmaps are either written to a directory or applied to a cluster.
+func (o *MirrorOptions) handleSignatures(context context.Context, signaturesByDigest map[string][][]byte) {
+	var client kubeclient.ConfigMapInterface
+	if !o.DryRun {
+		if o.ApplyReleaseImageSignature {
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go [hunk 15]
```diff
@@ -176,13 +223,123 @@ func (o *MirrorOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []
...
+			var err error
+			client, err = o.KubeClientFn()
+			if err != nil {
+				fmt.Fprintf(o.Out, "error: Can't apply manifests to cluster -%v\n", err)
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go [hunk 16]
```diff
@@ -213,6 +370,22 @@ func (o *MirrorOptions) Run() error {
...
+		if len(o.ReleaseImageSignatureToDir) == 0 {
+			if len(o.ToDir) == 0 {
+				return fmt.Errorf("if --to-dir and --apply-release-image-signature are not specified must use --release-image-signature-to-dir to specify a directory to export signature to")
+			}
+			o.ReleaseImageSignatureToDir = o.ToDir + configFilesBaseDir
+		}
+		if o.OverWrite {
+			return fmt.Errorf("--overwite is only valid when --apply-release-image-signature is specified")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go:388]: Changed in commit e3f50c1

File: pkg/verify/configmap.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+	"strings"
+
+	"sigs.k8s.io/yaml"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/openpgp"
+	"k8s.io/klog"
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go]: Fixed in commit e3f50c1

File: pkg/verify/configmap.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+	var cm *corev1.ConfigMap
+	cm, err := GetSignaturesAsConfigmap(digest, signatures)
+	if err != nil {
+		return nil, err
+	}
+	if cmData, err := yaml.Marshal(cm); err == nil {
+		return cmData, nil
+	} else {
+		return nil, fmt.Errorf("failed to create %s manifest: %v", digest, err)
+	}
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go]: Fixed in commit e3f50c1

File: pkg/verify/configmap.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+		cm.BinaryData[fmt.Sprintf("%s-%d", prefix, i+1)] = signatures[i]
+	}
+	return cm, nil
+}
+
+// GetSignaturesAsConfigmapBytes stores the given signatures as a configmap and
+// and returns configmap as bytes. Uses digestToKeyPrefix to replace colon with
+// dash when saving digest to configmap.
+func GetSignaturesAsConfigmapBytes(digest string, signatures [][]byte) ([]byte, error) {
+	var cm *corev1.ConfigMap
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go]: Fixed in commit e3f50c1

File: pkg/verify/configmap.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/verify/configmap.go]: Changed in commit e3f50c1

File: pkg/cli/admin/release/mirror.go [hunk 20]
```diff
@@ -20,6 +23,9 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
+	kubeclient "k8s.io/client-go/kubernetes/typed/core/v1"
```
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Fixed in commit e3f50c1

File: pkg/cli/admin/release/extract.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/cli/admin/release/extract.go:241]: Right, but you don't have any kind of loop in your code, that's handled where `TarEntryCallback` is being used. Thus you need to stop the execution when you encounter an error, if you don't your code will start panicing trying to read empty `raw`.

File: pkg/cli/admin/release/extract.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/cli/admin/release/extract.go:245]: See my reply.

File: pkg/cli/admin/release/extract.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/cli/admin/release/extract.go:250]: This was agreed to be fixed as a followup.

File: pkg/cli/admin/release/extract.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/cli/admin/release/extract.go:283]: Your suggestion still doesn't suggest the inverted flow. First errors, then information that nothing was found.

File: pkg/cli/admin/release/mirror.go [hunk 21]
```diff
@@ -213,6 +371,29 @@ func (o *MirrorOptions) Run() error {
...
+		if o.Overwrite {
+			return fmt.Errorf("--overwite is only valid when --apply-release-image-signature is specified")
+		}
+	}
+	return nil
+}
+
+func (o *MirrorOptions) Run() error {
+
+	if err := o.validate(); err != nil {
```
Comment (soltysh) [inline pkg/cli/admin/release/mirror.go:394]: You don't call validate here, this should live between these two lines above:
```go
kcmdutil.CheckErr(o.Complete(cmd, f, args))
kcmdutil.CheckErr(o.Run())
```
compare https://github.com/openshift/oc/blob/master/pkg/cli/extract/extract.go#L83-L87, for example.

File: pkg/cli/admin/release/extract.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:252]: All fixed

File: pkg/cli/admin/release/mirror.go (hunk 21 above)
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: All done

File: pkg/cli/admin/release/extract.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:241]: Sorry, this shouldn't have been so hard, had tunnel vision. Returns are in...

File: pkg/cli/admin/release/extract.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:252]: correct

File: pkg/cli/admin/release/extract.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:246]: Inserted return

File: pkg/cli/admin/release/mirror.go (hunk 21 above)
Comment (jottofar) [reply pkg/cli/admin/release/mirror.go]: Moved Validate per referenced pattern.

File: pkg/cli/admin/release/extract.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:241]: Inserted return statement

File: pkg/cli/admin/release/extract.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:246]: Inserted return statement

File: pkg/cli/admin/release/extract.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (wking) [reply pkg/cli/admin/release/extract.go:284]: I think the inverted flow thing is done per: https://github.com/openshift/oc/pull/343#discussion_r410451130  Can we mark this thread resolved?

[the remark this answers is not in the document]
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:284]: Changed to check for errors first

File: pkg/cli/admin/release/extract.go [hunk 22]
```diff
@@ -196,23 +210,86 @@ func (o *ExtractOptions) Run() error {
...
+							return true, nil
+						}
+						ms, err := manifest.ParseManifests(bytes.NewReader(raw))
+						if err != nil {
+							manifestErrs = append(manifestErrs, errors.Wrapf(err, "error parsing %s", hdr.Name))
+							return true, nil
+						}
+						for i := range ms {
+							ms[i].OriginalFilename = filepath.Base(hdr.Name)
+							src := fmt.Sprintf("the config map %s/%s", ms[i].Obj.GetNamespace(), ms[i].Obj.GetName())
```
Comment (wking) [inline pkg/cli/admin/release/extract.go:251]: Do we know it's a config map here?  I don't see you repeating [the CVO's `GVK` check][1].

[1]: https://github.com/openshift/cluster-version-operator/blob/914467c03b5bc25dbe4c9295c420fdb2598db821/pkg/cvo/cvo.go#L297

File: pkg/cli/admin/release/extract.go [hunk 23]
```diff
@@ -196,23 +210,86 @@ func (o *ExtractOptions) Run() error {
 				To:   dir,
 			},
 		}
+		var manifestErrs []error
 		found := false
 		opts.TarEntryCallback = func(hdr *tar.Header, _ extract.LayerInfo, r io.Reader) (bool, error) {
-			if hdr.Name != o.File {
+			if !o.ExtractManifests {
```
Comment (wking) [inline pkg/cli/admin/release/extract.go:216]: nit: you can avoid the double-negative of `if !a {b} else {c}` by inverting the blocks to `if a {c} else {b}`.  Not a big deal, but makes for easier reading for me, anyway.

File: pkg/cli/admin/release/extract.go (hunk 22 above)
Comment (soltysh) [reply pkg/cli/admin/release/extract.go:251]: I was suggesting using proper k8s encoding, as a follow-up. That will be the best.

File: pkg/cli/admin/release/extract.go (hunk 23 above)
Comment (jottofar) [reply pkg/cli/admin/release/extract.go:216]: I won't claim that one. Reused existing code and yeah, could have cleaned up, but was trying to minimize changes in the interest of making review easier (I've been dinged on that before) and generally minimizing changes.

**Comment by stbenjam** [comment]
Was this intended to be a breaking change? This broke the e2e-metal-ipi automation...


```
+(./04_setup_ironic.sh:68): main(): oc adm release mirror --insecure=true -a combined-pullsecret--gyUsq8fVYL --from registry.svc.ci.openshift.org/ci-op-hmxpdxpw/release@sha256:7d9cc5731a84efab3c59fe66a81a66ec1d3e386183b383f905d0efc71dd6f161 --to-release-image virthost.ostest.test.metalkube.org:5000/localimages/local-release-image:7d9cc5731a84efab3c59fe66a81a66ec1d3e386183b383f905d0efc71dd6f161 --to virthost.ostest.test.metalkube.org:5000/localimages/local-release-image
error: if --to-dir and --apply-release-image-signature are not specified, --release-image-signature-to-dir must be used to specify a directory to export the signature
```
